src/pool/srv_pool_scrub.c

```diff
@@ -0,0 +1,345 @@
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#define D_LOGFAC	DD_FAC(pool)
+
+#include <daos_srv/vos.h>
+#include <daos_srv/srv_csum.h>
+#include "srv_internal.h"
+
+#define C_TRACE(...) D_DEBUG(DB_CSUM, __VA_ARGS__)
+
+/** Hardcoded target interval for scrubber to complete */
+#define MSEC_IN_SEC 1000
+
+struct scrub_ctx {
+	/**
+	 * Pool
+	 **/
+	uuid_t			 pool_uuid;
+	daos_handle_t		 pool_hdl;
+	/** pool ULT schedule request. Used for yielding and sleeping */
+	struct sched_request	*req;
+	/** Number of checksums scrubbed by a single scrubbing iteration */
+	daos_size_t		 pool_csums_scrubbed;
+
+	/**
+	 * Container
+	 **/
+	struct ds_cont_child	*cur_cont;
+
+	/** Number of msec between checksum calculations */
+	daos_size_t		 msec_between_calcs;
+	/** Target number of seconds to scrub entire pool */
+	daos_size_t		 interval_sec;
+};
+
+static void
+sc_reset(struct scrub_ctx *ctx)
+{
+	ctx->pool_csums_scrubbed = 0;
+}
+
+/**
+ * The following 3 functions will be replaced by pool/container properties,
+ * but for now, to make testing easier using environment variables to configure.
+ */
+static bool
+scrubbing_is_enabled()
+{
+	char *enabled = getenv("DAOS_CSUM_SCRUB");
+
+	return enabled != NULL && strncmp(enabled, "ON", strlen("ON")) == 0;
+}
+
+static uint64_t
+interval_sec()
+{
+	char *sec = getenv("DAOS_CSUM_SCRUB_INTERVAL_SEC");
+
+	return sec != NULL ? atoll(sec) : 7 * 24 * 60 * 60;
+}
+
+static uint64_t
+between_scrub_sec()
+{
+	char *sec = getenv("DAOS_CSUM_SCRUB_PAUSE_SEC");
+
+	return sec != NULL ? atoll(sec) : 24 * 60 * 60;
+}
+
+/** vos_iter_cb_t */
+static int
+obj_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		  vos_iter_type_t type, vos_iter_param_t *param,
+		  void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	struct daos_csummer	*csummer = ctx->cur_cont->sc_csummer;
+
+	if (!(type == VOS_ITER_RECX || type == VOS_ITER_SINGLE))
+		return 0;
+
+	C_TRACE("Scrubbing akey: "DF_KEY", type: %s, record size: "
+			DF_U64", extent: "DF_RECX", csum_size: %d\n",
+		DP_KEY(&param->ip_akey),
+		(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
+		entry->ie_rsize,
+		DP_RECX(entry->ie_orig_recx),
+		daos_csummer_get_csum_len(csummer)
+	);
+
+	/** TODO - implement the actual scrubbing of the checksum by fetching
+	 * the data, calculating a new checksum, and comparing to the stored
+	 * checksum.
+	 */
+	ctx->pool_csums_scrubbed++;
+
+	if (ctx->msec_between_calcs == 0) {
+		C_TRACE("Yield after data scrub\n");
+		dss_ult_yield(ctx->req);
+	} else {
+		C_TRACE("Sleeping after data scrub for "DF_U64" msec\n",
+			ctx->msec_between_calcs);
+		sched_req_sleep(ctx->req, ctx->msec_between_calcs);
+	}
```

Comment:
Once yield in iterate callback, Inform vos_iterate() to do re-probe by VOS_ITER_CB_YIELD flag.